### ISSUE TITLE
Improving quality with 8bit?

### ISSUE THREAD

**Issue by neuhaus** [ggerganov/llama.cpp#53]
I can achieve around 1 token per second on a Ryzen 7 3700X on Linux with the 65B model and 4bit quantization.

If we use 8bit instead, would it run faster? I have 128GB RAM. Is 8bit already supported?
```
$ ./main -m models/65B/ggml-model-q4_0.bin -t 8 -n 128
main: mem per token = 70897348 bytes
main:     load time = 14010.35 ms
main:   sample time =   335.09 ms
main:  predict time = 140527.48 ms / 1089.36 ms per token
main:    total time = 157951.48 ms
```

**Comment by gjmulder** [comment]
I tried the intermediate fp16 and could get the model to run in 122GB of resident memory. With a Ryzen 1950X 16 Core CPU I'm generating tokens at about one per three seconds.

**Comment by gjmulder** [comment]
`./main -m ./models/65B/ggml-model-f16.bin -t 16 -n 128`

**Comment by neuhaus** [comment]
OK i tried it with the fp16 model too, it only swapped a little bit (i have an 8-core Ryzen 7 3700X and 128GB RAM):
```
$ ./main -m models/65B/ggml-model-f16.bin -t 8 -n 128
main: mem per token = 70897348 bytes
main:     load time = 71429.04 ms
main:   sample time =   324.53 ms
main:  predict time = 402116.09 ms / 3117.18 ms per token
main:    total time = 483291.78 ms
```
I also tried using `-t 16` (for take advantage of multithreading) but it ended up being slightly slower. I'm still hoping that 8bit might be faster than 4bit...?

**Comment by neuhaus** [comment]
As of now the "quantize" can only do 4bit.

**Comment by gjmulder** [comment]
122GB.

What would be interesting is to benchmark quality versus memory size, i.e. does say a fp16 13B model generate better output than a fp4 60GB model?

@apollotsantos are you in Lisboa? I'm in Carcavelos.

**Comment by ggerganov** [comment]
No 8-bit support atm, but can be added similar to [4-bit](https://github.com/ggerganov/ggml/pull/27)
I expect it will be slower, because it will increase memory traffic. But it also depends on how efficient the SIMD is implemented

**Comment by neuhaus** [comment]
I noticed a significant quality increase going from 13B to 30B but 65B is too slow on this CPU to use it much.

**Comment by gjmulder** [comment]
> @gjmulder actually not. I'm in Brazil

Lucky you!

**Comment by gjmulder** [comment]
This issue is perhaps misnamed, now, as 8bit will likely improve _quality_ over 4bit but not _performance_.

In summary:

- Inference performance: 4bit > 8bit > fp16 (as the code looks to be primarily memory-bound, with only a 50% performance increase from going from 8 cores to 16 cores on my 16 core / 32 hyperthread Ryzen 1650X.
- Precision quality: fp16 > 8bit > 4bit (as more precision means improved inference quality)
- Scaling quality: 65B > 30B > 13B > 7B (scaling of models improves inference quality significantly)

Which led me to wonder where the sweet spots are among two parameters for a given memory footprint?

Once the model is able to be loaded once and called repeatedly (issue #23) and the python bindings are merged (issue #82 and https://github.com/thomasantony/llama.cpp/tree/feature/pybind), I can test all the permutations against say the [SQuAD](https://rajpurkar.github.io/SQuAD-explorer/) benchmark and we can understand the impact of quantization versus model size.

**Comment by AGSaidi** [comment]
Arm has [SMMLA instructions](https://developer.arm.com/documentation/ddi0596/2020-12/SIMD-FP-Instructions/SMMLA--vector---Signed-8-bit-integer-matrix-multiply-accumulate--vector--) which for newer arm targets should give another 4x over fp16.

**Comment by MarkSchmidty** [comment]
> 122GB.
> 
> What would be interesting is to benchmark quality versus memory size, i.e. does say a fp16 13B model generate better output than a int4 60GB model?
> 
> @apollotsantos are you in Lisboa? I'm in Carcavelos.

The answer is no, At around 20B parameters you only need 3 bits to get about the same performance quality as the same 20B parameter model in uncompressed fp16. After that, for each 4x more parameters you can drop a bit off while still getting close to 16bit quality.

So an 80B parameter model would have around the same quality in 2bit as in 16bit and a 320B parameter model would have around the same quality in 1bit as in 16bit. Beyond 1bit quantization can be achieved through various methods, such as re-using bins of bits from non-connected layers, which are only applicable to massive models and which will only maintain output quality for ~1T+ parameter models.

I'm not going to list every source for this. But these papers are a good start:
[GPTQ: Accurate Post-Training Quantization for Generative Pre-trained Transformers | Oct 2022](https://arxiv.org/abs/2210.17323)
[The case for 4-bit precision: k-bit Inference Scaling Laws | Dec 2022 - Updated Feb 2023](https://arxiv.org/abs/2212.09720)
[SparseGPT: Massive Language Models Can Be Accurately Pruned in One-Shot | Jan 2023](https://arxiv.org/abs/2301.00774)

**Comment by MarkSchmidty** [comment]
>Which led me to wonder where the sweet spots are among two parameters for a given memory footprint?

13B appears to have negligible quality difference at 3-bit.

So you'll want to 13B-65B in 3-bit to save memory and run faster for effectively the same quality output, once it is implemented.

For 7B 4bit is practically always best. If you really want to run it in 4GB of memory then 30bit will make it fit at a reduced quality, but not so much as to make it unusable, especially with finetuning.

Some interesting use cases for 4GB inference including running at near native speeds in the browser with WebAssembly and running on the very popular Raspberry Pi 4GB. :)

**Comment by j-f1** [comment]
> Some interesting use cases for 4GB inference including running at near native speeds fully in a web browser on any device with WebAssembly and running on the very popular Raspberry Pi 4GB. :)

Also newish iPhones which allow up to 4080MB of memory use with the “Increased Memory Limit” entitlement!

**Comment by riverzhou** [comment]
Waiting for int8 quantization....
